net/keepalived/Makefile

```diff
@@ -274,4 +274,33 @@ endif
 
 endef
 
+define Package/keepalived-sync
+  SECTION:=net
+  CATEGORY:=Network
+  TITLE:=Keepalived Master and Backup Synchronization
+  DEPENDS:= +keepalived +rsync +inotifywait
+endef
+
+define Package/keepalived-sync/description
+ Keepalived Master Backup sync files
+endef
+
+define Package/keepalived-sync/install
+	$(INSTALL_DIR) $(1)/etc/init.d \
+		$(1)/etc/keepalived/scripts \
+		$(1)/etc/keepalived/keys \
```

Comment:
Why do you create this directory. We are not installing anything there.

Reply:
We created these directories to keep all scripts and keys in this location. From the Luci application, we are allowing the upload and selection of scripts/keys from this directory only in vrrp_script and peer configuration respectively, also as /etc/keepalived/ is made part of conf files, It would also be backed up with config backup.

rsync.sh is installed in /etc/keepalived/scripts.

Reply:
>> Where is the rsync.sh used at all?

rsync needs UCI configuration of vrrp_script and track_script.

First Its configured as vrrp_script and then track_script
Now when we configure vrrp_instance, we add rsync script in a list of track_script configuration

We did not add any cron job/service because it only has to be executed when keepalived is running and as keepalived already provides a function to periodically execute the script, we leveraged that feature

Here is an example working UCI config from the backup node.

```sh
config globals 'globals'

config ipaddress
	option name 'eth0_ha'
	option address '192.168.111.101'
	option device 'eth0'
	option label_suffix 'ha'

config ipaddress
	option name 'eth3_ha'
	option address '192.168.5.1'
	option device 'eth3'
	option label_suffix 'ha'

config vrrp_script
	option name 'sync'
	option script '/etc/keepalived/scripts/rsync.sh'
	option interval '60'
	option weight '100'

config static_ipaddress
	list address 'eth0_ha'
	list address 'eth3_ha'

config track_script
	option name 'ha_sync'
	option value 'sync'
	option weight '100'

config track_interface
	option name 'eth0_ha'
	option value 'eth0'
	option weight '100'

config track_interface
	option name 'eth3_ha'
	option value 'eth3'
	option weight '100'

config peer
	option name 'master'
	option address '100.100.100.1'
	option sync '1'
	option sync_mode 'receive'
	option sync_dir '/usr/share/keepalived/rsync'
	option ssh_pubkey 'authorized_key_string'

config vrrp_instance
	option name 'backup'
	option state 'BACKUP'
	option interface 'eth1'
	option virtual_router_id '100'
	option priority '50'
	option advert_int '1'
	option nopreempt '0'
	list virtual_ipaddress 'eth0_ha'
	list virtual_ipaddress 'eth3_ha'
	option unicast_src_ip '100.100.100.2'
	list unicast_peer 'master'
	option auth_type 'PASS'
	option auth_pass 'admin'
	list track_interface 'eth0_ha'
	list track_interface 'eth3_ha'
	list track_script 'ha_sync'
```
rsync works differently based on if the node is master/backup. If the node is master, it sends configuration files to the backup node. If the node is back up it checks and updates authorized_keys.